File: javascript/array-hasing/longest-consecutive-sequence.js
// Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.

// You must write an algorithm that runs in O(n) time.

function longestConsecutive(nums) {
    let longest = 0
    const numSet = new Set(nums)

    for(const num of numSet) {
        // find the lowest or the start number in one of the sequences
        if (numSet.has(num - 1)) continue
        // save the start of one sequence and his length
        let [currentNum, len] = [num, 1]
        // search for elements that follow in the corresponding sequence
        while(numSet.has(currentNum+1)) {
            // move to the next potential number
            currentNum++
            // add length
            len++
        }
    
        longest = Math.max(longest, len)
    }

    return longest
}

// testing 
console.log(longestConsecutive([100,4,200,1,3,2]))
// expected output: 4

File: python/array-hashing/longest_consecutive_sequence.py
# Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.

# You must write an algorithm that runs in O(n) time.

def longestSequence(nums: list) -> int:
    longest = 0
    numSet = set(nums)

    for num in numSet:
        # find the start of one sequence
        if num-1 not in numSet:
            length = 1
            # search for the next numbers in this sequence
            while num + 1 in numSet:
                num += 1
                length += 1
            
            longest = max(longest, length)
    
    return longest

# testing
print(longestSequence([100,4,200,1,3,2]))
# expected output: 4